fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1263,212 @@ class SamOrder(enum.Enum):
     Coordinate = "coordinate"  #: coordinate sorted
     QueryName = "queryname"  #: queryname sorted
     Unknown = "unknown"  # Unknown SAM / BAM / CRAM sort order
+
+
+@dataclass(frozen=True)
+class SecondaryAlignment:
+    """A secondary alignment as encoded in one part of the `XA` or `XB` tag value on a SAM record.
+
+    Format of a single secondary alignment in the `XA` tag (`,`-delimited):
+
+    ```text
+    chr,<orientation><position>,cigar,NM
+    ```
+
+    Full example of an `XA` tag value (`;`-delimited):
+
+    ```text
+    XA:Z:chr9,-104599381,49M,4;chr3,+170653467,49M,4;chr12,+46991828,49M,5;
+    ```
+
+    Format of a single secondary alignment in the `XB` tag (`,`-delimited):
+
+    ```text
+    chr,<orientation><position>,cigar,NM,AS,MapQ
+    ```
+
+    Full example of an `XB` tag value (`;`-delimited):
+
+    ```text
+    XB:Z:chr9,-104599381,49M,4,0,30;chr3,+170653467,49M,4,0,20;chr12,+46991828,49M,5,0,10;
+    ```
+
+    Args:
+        reference_name: The reference sequence name.
+        reference_start: The 0-based start position of the alignment.
+        is_forward: If the alignment is in the forward orientation or not.
+        cigar: The Cigar sequence representing the alignment.
+        edit_distance: The number of mismatches between the query and the target.
+        alignment_score: The aligner-reported alignment score, if available.
+        mapq: The aligner-reported probability of an incorrect mapping, if available.
+
+    See:
+        - [BWA User Manual](https://bio-bwa.sourceforge.net/bwa.shtml)
+        - [https://github.com/lh3/bwa/pull/292](https://github.com/lh3/bwa/pull/292)
+        - [https://github.com/lh3/bwa/pull/293](https://github.com/lh3/bwa/pull/293)
+        - [https://github.com/lh3/bwa/pull/355](https://github.com/lh3/bwa/pull/355)
+    """
+
+    reference_name: str
+    reference_start: int
+    is_forward: bool
+    cigar: Cigar
+    edit_distance: int
+    alignment_score: Optional[int] = None
+    mapq: Optional[int] = None
+
+    def __post_init__(self) -> None:
+        """Perform post-initialization validation on this dataclass."""
+        if self.reference_start < 0:
+            raise ValueError(f"Start cannot be <0! Found: {self.reference_start}")
+        if self.edit_distance < 0:
+            raise ValueError(f"Edit distance cannot be <0! Found: {self.edit_distance}")
+        if self.alignment_score is not None and self.alignment_score < 0:
+            raise ValueError(f"Alignment score cannot be <0! Found: {self.alignment_score}")
+        if self.mapq is not None and self.mapq < 0:
+            raise ValueError(f"Mapping quality cannot be <0! Found: {self.mapq}")
```

Comment:
**suggestion** I think it'd be helpful to collect all possible error messages and raise at the end.

It'd also be helpful to include the query name - this info isn't in the scope of the dataclass constructor, so I'd wrap the constructions in the various classmethods below in a try-except to include the query name when constructing from AlignedSegment


```suggestion
        errs: list[str] = []
        if self.reference_start < 0:
            errs.append(f"Start cannot be <0! Found: {self.reference_start}")
        if self.edit_distance < 0:
            errs.append(f"Edit distance cannot be <0! Found: {self.edit_distance}")
        if self.alignment_score is not None and self.alignment_score < 0:
            errs.append(f"Alignment score cannot be <0! Found: {self.alignment_score}")
        if self.mapq is not None and self.mapq < 0:
            errs.append(f"Mapping quality cannot be <0! Found: {self.mapq}")
            
        if errs:
            raise ValueError("\n".join(errs))
```